src/gtest/test_noteencryption.cpp

```diff
@@ -19,6 +20,99 @@ class TestNoteDecryption : public ZCNoteDecryption {
     }
 };
 
+TEST(noteencryption, NotePlaintext)
+{
+    using namespace libzcash;
+    auto sk = SaplingSpendingKey(uint256()).expanded_spending_key();
+    auto vk = sk.full_viewing_key();
+    auto ivk = vk.in_viewing_key();
+    SaplingPaymentAddress addr = *ivk.address({0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0});
+
+    std::array<unsigned char, ZC_MEMO_SIZE> memo;
+    for (size_t i = 0; i < ZC_MEMO_SIZE; i++) {
+        // Fill the message with dummy data
+        memo[i] = (unsigned char) i;
+    }
+
+    SaplingNote note(addr, 39393);
+    SaplingNotePlaintext pt(note, memo);
+
+    auto res = pt.encrypt(addr.pk_d);
+    if (!res) {
+        FAIL();
+    }
+
+    auto enc = res.get();
+
+    auto ct = enc.first;
+    auto encryptor = enc.second;
+    auto epk = encryptor.get_epk();
+
+    // Try to decrypt
+    auto foo = SaplingNotePlaintext::decrypt(
+        ct,
+        ivk,
+        epk
+    );
+
+    if (!foo) {
+        FAIL();
+    }
+
+    auto bar = foo.get();
+
+    ASSERT_TRUE(bar.value() == pt.value());
+    ASSERT_TRUE(bar.memo() == pt.memo());
+    ASSERT_TRUE(bar.d == pt.d);
+    ASSERT_TRUE(bar.rcm == pt.rcm);
+
+    auto foobar = bar.note(ivk);
+
+    if (!foobar) {
+        FAIL();
+    }
+
+    auto new_note = foobar.get();
+
+    ASSERT_TRUE(note.value() == new_note.value());
+    ASSERT_TRUE(note.d == new_note.d);
+    ASSERT_TRUE(note.pk_d == new_note.pk_d);
+    ASSERT_TRUE(note.r == new_note.r);
+    ASSERT_TRUE(note.cm() == new_note.cm());
+
+    SaplingOutgoingPlaintext out_pt;
+    out_pt.pk_d = note.pk_d;
+    out_pt.esk = encryptor.get_esk();
+
+    auto ovk = random_uint256();
+    auto cv = random_uint256();
+    auto cm = random_uint256();
+
+    auto out_ct = out_pt.encrypt(
+        ovk,
+        cv,
+        cm,
+        encryptor
+    );
+
+    auto decrypted_out_ct = out_pt.decrypt(
+        out_ct,
+        ovk,
+        cv,
+        cm,
+        encryptor.get_epk()
+    );
+
+    if (!decrypted_out_ct) {
+        FAIL();
+    }
+
+    auto decrypted_out_ct_unwrapped = decrypted_out_ct.get();
+
+    ASSERT_TRUE(decrypted_out_ct_unwrapped.pk_d == out_pt.pk_d);
+    ASSERT_TRUE(decrypted_out_ct_unwrapped.esk == out_pt.esk);
```

Comment:
I don't see here a testing of actually decrypting the outgoing note correctly, and not just the "metadata" (do we have better name for it?) (esk,pkd)

Reply:
Ahh I guess the metadata is what we're calling outgoing plaintext,  we might need to explain that somewhere as at first read outgoing plaintext seems to be everything that's going out, including the note plaintext @daira

Reply:
In pairing with @arielgabizon we identified we need to test the sender decrypting the note ciphertext, using the values obtained through decrypting the "metadata".

Reply:
@arielgabizon I added decryption using full viewing key per spec 4.17.3 and updated the test.

Reply:
Nice job! Except perhaps the method names having 'UsingFullViewingKey'.. the method name suggests
you're doing the whole process - getting (esk,pkd) and decrypting note in that method;
whereas you're just doing the second part after getting (esk,pkd),
so I'd either refactor to get a method that does the whole thing under this name,
or call it 'UsingOutgoingPlaintext'

Reply:
@arielgabizon I've updated to just call it decrypt so we use function overloading for now.    'UsingOutgoingPlaintext' implies we pass in a SaplingOutgoingPlaintext object when in fact we are just passing in some components. I think we can rename things at a later date.  If this is good, please ack review.